### PR TITLE
feat: Add nativeStability.acknowledgeReports API

### PR DESCRIPTION
**Pull request by hlwarriner**
This web API enables a web app to acknowledge native stability events as already handled so they are no longer returned by getPendingReports(). Acknowledged event UUIDs are persisted to disk so that the behavior survives across Cobalt app lifecycles; storage is managed by the NativeStabilityManager. More context can be found in the "High level" and "Storage 3P MVP" design docs.

To manage the scope of this PR, pruning of the persisted event UUIDs - discussed in the design doc - is left as a quick follow-up.

A stub Mojo binding is provided for the native stability interface for non-Evergreen platforms so that web requests on these platforms will result in rejected promises instead of any actual state changes.

A couple other small improvements are included:
* An existing unit test case is updated to remove noisy logs
* Handling of Mojo disconnection is added to the renderer impl.

Issue: 528362453

### CODE REVIEW

**Comment by hlwarriner** [comment]
CC @AlecsisDuarte @andrewsavage1

File: third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc [hunk 1]
```diff
@@ -98,11 +123,26 @@ void H5vccNativeStability::EnsureReceiverIsBound() {
...
+  remote_native_stability_.reset();
+  HeapHashSet<Member<ScriptPromiseResolverBase>> pending_promises;
+  // Script may execute during a call to Reject(). Swap these sets to prevent
+  // concurrent modification.
+  ongoing_requests_.swap(pending_promises);
+  for (auto& resolver : pending_promises) {
+    resolver->Reject("Mojo connection error.");
+  }
+  ongoing_requests_.clear();
 }
```
[the remark this answers is not in the document]
Comment (hlwarriner) [reply third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc:139]: done

**Comment by hlwarriner** [comment]
Sending a reminder to please take a look at this, @y4vor @vincentmo @alunev.

File: cobalt/browser/h5vcc_native_stability/native_stability_manager_unittest.cc [hunk 2]
```diff
@@ -255,4 +274,79 @@ TEST_F(NativeStabilityManagerTest, GetPendingReportsIgnoresUnknownReportType) {
   run_loop.Run();
 }
 
+TEST_F(NativeStabilityManagerTest,
+       AcknowledgedReportsFilteredByGetPendingReports) {
```
Comment (AlecsisDuarte) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager_unittest.cc:278]: To make these unit tests more robust, should we add the following unit tests:
 - empty UUID strings or duplicates?
 - and a corrupted or non-array `acked_event_uuids.json`?

File: third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc [hunk 3]
```diff
@@ -87,6 +106,12 @@ void H5vccNativeStability::OnGetPendingReports(
   resolver->Resolve(std::move(result));
 }
 
+void H5vccNativeStability::OnAcknowledgeReports(
```
Comment (AlecsisDuarte) [inline third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc:109]: Should we guard against race conditions with `OnConnectionError` by verifying `if (ongoing_requests_.Contains(resolver)) { return; }` before `erase`?

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc [hunk 4]
```diff
@@ -36,6 +55,79 @@ mojom::BaseReportDataPtr CreateBaseReportData(
...
+  std::optional<base::Value::List> parsed_list =
+      base::JSONReader::ReadList(file_content);
+  if (!parsed_list) {
+    LOG(WARNING) << "Failed to parse acked UUIDs JSON list in: "
+                 << file_path.value();
+    return acked_uuids;
+  }
+
+  for (const auto& item : *parsed_list) {
+    if (item.is_string()) {
```
Comment (AlecsisDuarte) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:81]: Super-nit: You could easily reduce branching here by wrapping the warning and continuing instead of an if/else branching 

```cc
if (!item.is_string()} {
     LOG(WARNING) << "Skipping non-string item in acked UUIDs list: "
                   << file_path.value();
     continue;
}
acked_uuids.insert(item.GetString());
```

**Review by AlecsisDuarte — APPROVED** [review]
LGTM! Left non-blocking comments.

File: third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc (hunk 3 above)
Comment (hlwarriner) [reply third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc:109]: Good question, thanks for raising this. It should be ok as is, though, for two reasons:
1. `HeapHashSet` inherits `erase()` from `blink::HashSet::erase()`, which is safe and just "does nothing if the value is not found"
2. The `OnGetPendingReports`, `OnAcknowledgeReports`, and `OnConnectionError` callbacks should all run on the same single-threaded task runner. So they execute sequentially.

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc (hunk 4 above)
Comment (hlwarriner) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:81]: Thanks for the feedback. I do often like using early exits/returns to avoid overly nesting the base case but I think here the if/else works fine: the if and else branches are very short and the if/else clearly expresses the binary choice.

File: cobalt/browser/h5vcc_native_stability/native_stability_manager_unittest.cc (hunk 2 above)
Comment (hlwarriner) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager_unittest.cc:278]: Good idea. I'll add these.

For the corrupted ack file, I was trying to keep the unit tests more "black box" and not interact with implementation details like the ack file in the test setup. But I think it's an important enough edge case to test that it's worth compromising on that principle.

**Review by AlecsisDuarte — APPROVED** [review]
Great, thanks!

File: third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc [hunk 5]
```diff
@@ -87,6 +106,12 @@ void H5vccNativeStability::OnGetPendingReports(
...
+    ScriptPromiseResolver<IDLUndefined>* resolver) {
+  ongoing_requests_.erase(resolver);
+  resolver->Resolve();
+}
+
 void H5vccNativeStability::EnsureReceiverIsBound() {
   DCHECK(GetExecutionContext());
```
Comment (alunev) [inline third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc:116]: I think this is compiled out in release and runtime check is needed like in https://github.com/youtube/cobalt/blob/main/third_party/blink/renderer/modules/cobalt/h5vcc_platform_service/h_5_vcc_platform_service.cc#L214

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc [hunk 6]
```diff
@@ -143,4 +269,44 @@ void NativeStabilityManager::GetPendingReportsOnTaskRunner(
...
+  for (const auto& uuid : native_stability_event_uuids) {
+    if (!uuid.empty() && acked_uuids.insert(uuid).second) {
+      changed = true;
+    }
+  }
+
+  if (changed) {
+    // It's simpler and safer to just overwrite the file with previously +
+    // currently acknowledged event UUIDs, rather than append to it.
+    WriteAckedUuidsToDisk(file_path, acked_uuids);
```
Comment (alunev) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:305]: nit: return value is ignored here, let's log it to avoid silent failure?

**Review by alunev — APPROVED** [review]

File: third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc (hunk 5 above)
Comment (hlwarriner) [reply third_party/blink/renderer/modules/cobalt/h5vcc_native_stability/h_5_vcc_native_stability.cc:116]: Yeah, it is. I think it's a good discussion to have about whether we prefer
1) ` DCHECK(GetExecutionContext())`
2)  `CHECK(GetExecutionContext())`
3) `if (!GetExecutionContext()) return`
here.

But I think we should have that discussion independently from this PR.

Most other Cobalt modules are using  `DCHECK(GetExecutionContext())`, as well. So doing `if (!GetExecutionContext()) return` here would be a departure from that convention.

`H5vccPlatformService::OnDataReceived` is different because it runs asynchronously to handle incoming Mojo events from the browser, at which point it's more likely the page could be gone. (As opposed to running synchronously in response to a web API call from the web client, when the page should be valid.)

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc (hunk 6 above)
Comment (hlwarriner) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:305]: Good point. I changed the return type to `void` and made sure every failure case in `WriteAckedUuidsToDisk()` includes a log.

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.h [hunk 7]
```diff
@@ -57,12 +60,26 @@ class NativeStabilityManager {
...
+  // As with GetPendingReports(), disk I/O is offloaded to a background
+  // ThreadPool task runner and callbacks are posted back to the UI sequence.
+  void AcknowledgeReports(std::vector<std::string> native_stability_event_uuids,
+                          base::OnceClosure callback);
+
   using GetExtensionCallback =
       base::RepeatingCallback<const void*(const char*)>;
 
-  // Inject a custom Starboard extension getter callback for testing.
+  // Injects a custom Starboard extension getter callback for testing.
```
Comment (vincentmo) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager.h:84]: nit: can we move all testing only methods to the bottom of the method declaration section?

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.h [hunk 8]
```diff
@@ -77,11 +94,18 @@ class NativeStabilityManager {
       base::OnceCallback<void(std::vector<mojom::NativeStabilityReportPtr>)>
           callback);
 
+  void AcknowledgeReportsOnTaskRunner(
```
Comment (vincentmo) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager.h:104]: When should this be used over AcknowledgeReports? Should one of them be declared private instead?

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc [hunk 9]
```diff
@@ -143,4 +268,44 @@ void NativeStabilityManager::GetPendingReportsOnTaskRunner(
...
+      base::BindOnce(&NativeStabilityManager::AcknowledgeReportsOnTaskRunner,
+                     base::Unretained(this),
+                     std::move(native_stability_event_uuids),
+                     base::BindPostTaskToCurrentDefault(std::move(callback))));
+}
+
+void NativeStabilityManager::AcknowledgeReportsOnTaskRunner(
+    std::vector<std::string> native_stability_event_uuids,
+    base::OnceClosure callback) {
+  base::FilePath file_path = GetAckedUuidsFilePath();
```
Comment (vincentmo) [inline cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:291]: I'm not sure if this scenario is valid, but is it possible for there to be multiple instances of cobalt running on a device (for example, YouTube TV and YouTube)? And if so, would they share the same file?
I'm wondering if we need to globally lock the file during the process to prevent concurrent process access, among other things...

**Review by vincentmo — APPROVED** [review]
See my question about multi process access, otherwise, lgtm.

**Review by y4vor — APPROVED** [review]

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.cc (hunk 9 above)
Comment (y4vor) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:291]: oh, yeah this is possible. @hlwarriner are you partitioning the file storage?

Comment (hlwarriner) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager.cc:291]: This is a good point.

We are not specifically partitioning this file so we'll just get whatever partitioning is provided by `kSbSystemPathCacheDirectory`. From what I can tell Evergreen platforms may, but are not guaranteed, to partition `kSbSystemPathCacheDirectory` for different applications.

Given this I think we can say that:
1. We can guarantee file integrity.
2. But we may have read-modify-write data races.

Since 2 should not occur very often and would not be disastrous if it did, I'll just add a todo comment for now - for us to consider partitioning the file.

For 1, `base::ImportantFileWriter::WriteFileAtomically()` should provide safe, atomic writes to the file. On POSIX platforms the implementation first creates a temporary file *with a unique name* with the new contents, and then replaces the file with the temporary file via `rename()`, which is itself guaranteed to be atomic.

For 2, we update the ack file by reading the existing file, modifying the set of UUIDs in memory by adding new ones from the current request, and then writing the updated set back to the ack file. It should be possible for something like the following to happen:
* Process A reads acked_event_uuids.json (sees ["uuid-1"]).
* Process B reads acked_event_uuids.json (sees ["uuid-1"]).
* Process A appends "uuid-2" and writes ["uuid-1", "uuid-2"] via temp file + rename().
* Process B appends "uuid-3" and writes ["uuid-1", "uuid-3"] via temp file + rename().

Here, Process B would "win" with the last write and "uuid-2" would be lost.

File: cobalt/browser/h5vcc_native_stability/native_stability_manager.h (hunk 7 above)
Comment (hlwarriner) [reply cobalt/browser/h5vcc_native_stability/native_stability_manager.h:84]: The `*ForTesting` methods already are grouped together at the bottom of the `public` declaration section.
